File: 03-data-structures/01-arrays/practice/solutions/07-move-zeroes.py
def move_zeroes(nums: list[int]) -> None:
    """
    LeetCode 283: Move Zeroes

    Problem: Move all zeros to the end of the array while maintaining
    the order of non-zero elements. Do this in-place.

    Approach: Two-pointer technique.
    - write: position for next non-zero
    - read: scans through array

    Time: O(n) - single pass for non-zeros + second pass for zeros
    Space: O(1) - in-place modification
    """
    write = 0  # Position for next non-zero

    # First pass: move all non-zeros to the front
    for read in range(len(nums)):
        if nums[read] != 0:
            nums[write] = nums[read]
            write += 1

    # Second pass: fill remaining positions with zeros
    for i in range(write, len(nums)):
        nums[i] = 0


def move_zeroes_swap(nums: list[int]) -> None:
    """
    Alternative: single pass with swapping.
    More intuitive: swap each non-zero with the write position.
    """
    write = 0  # Position for next non-zero

    for read in range(len(nums)):
        if nums[read] != 0:
            # Swap nums[read] and nums[write]
            nums[write], nums[read] = nums[read], nums[write]
            write += 1


# Test cases
if __name__ == "__main__":
    arr1 = [0, 1, 0, 3, 12]
    move_zeroes(arr1)
    print(arr1)  # [1, 3, 12, 0, 0]

    arr2 = [0, 0, 0, 1]
    move_zeroes(arr2)
    print(arr2)  # [1, 0, 0, 0]

    arr3 = [1, 2, 3]
    move_zeroes(arr3)
    print(arr3)  # [1, 2, 3]

File: 03-data-structures/01-arrays/practice/solutions/07-move-zeroes.ts
/**
 * LeetCode 283: Move Zeroes
 * 
 * Problem: Move all zeros to the end of the array while maintaining
 * the order of non-zero elements. Do this in-place.
 * 
 * Approach: Two-pointer technique similar to remove element.
 * - write: position for next non-zero
 * - read: scans through array
 * 
 * Time: O(n) - single pass
 * Space: O(1) - in-place modification
 */

function moveZeroes(nums: number[]): void {
    let write = 0;  // Position for next non-zero
    
    // First pass: move all non-zeros to the front
    for (let read = 0; read < nums.length; read++) {
        if (nums[read] !== 0) {
            nums[write] = nums[read];
            write++;
        }
    }
    
    // Second pass: fill remaining positions with zeros
    for (let i = write; i < nums.length; i++) {
        nums[i] = 0;
    }
}

// Alternative: single pass with swapping
function moveZeroesSwap(nums: number[]): void {
    let write = 0;  // Position for next non-zero
    
    for (let read = 0; read < nums.length; read++) {
        if (nums[read] !== 0) {
            // Swap nums[read] and nums[write]
            const temp = nums[write];
            nums[write] = nums[read];
            nums[read] = temp;
            write++;
        }
    }
}

// Test cases
const arr1 = [0, 1, 0, 3, 12];
moveZeroes(arr1);
console.log(arr1);  // [1, 3, 12, 0, 0]

const arr2 = [0, 0, 0, 1];
moveZeroes(arr2);
console.log(arr2);  // [1, 0, 0, 0]

const arr3 = [1, 2, 3];
moveZeroes(arr3);
console.log(arr3);  // [1, 2, 3] (no zeros, unchanged)

/**
 * Visual trace for [0, 1, 0, 3, 12]:
 * 
 * First pass (moving non-zeros):
 * read=0: nums[0]=0, skip
 * read=1: nums[1]=1, write to nums[0], nums=[1, 1, 0, 3, 12], write=1
 * read=2: nums[2]=0, skip
 * read=3: nums[3]=3, write to nums[1], nums=[1, 3, 0, 3, 12], write=2
 * read=4: nums[4]=12, write to nums[2], nums=[1, 3, 12, 3, 12], write=3
 * 
 * Second pass (filling zeros):
 * i=3: nums[3]=0, nums=[1, 3, 12, 0, 12]
 * i=4: nums[4]=0, nums=[1, 3, 12, 0, 0]
 * 
 * Final: [1, 3, 12, 0, 0]
 */